### ISSUE TITLE
Namespace prevents customAutoEscapedFunctions from being recognized

### ISSUE THREAD

**Issue by paulschreiber** [WordPress-Coding-Standards/WordPress-Coding-Standards#1176]
PHPCS 3.1, WPCS 0.13.0, PHP 5.6.30

```

**Comment by jrfnl** [comment]
@paulschreiber Without looking at the code in detail, the property expects to receive function names, not class or namespace names. Try passing it `sanitize_meta_css` instead.

**Comment by paulschreiber** [comment]
@jrfnl I'd tried that. When `customAutoEscapedFunctions` has `value="sanitize_meta_css"`, PHPCS shows the same error "Expected next thing to be an escaping function" as above.

**Comment by GaryJones** [comment]
I have come across issues with `\` scoped items, but under a slightly different context (and I can't find the relevant issue).

**Comment by jrfnl** [comment]
Ok, this will need further investigation. Related to #764

**Comment by GaryJones** [comment]
#933 is the one I was thinking of.

**Comment by paulschreiber** [comment]
This looks just like #933.

**Comment by dingo-d** [comment]
This doesn't seem to be fixed, or I'm doing something wrong. I have a code like

```php
echo Components::render(
	'modal',
	array_merge(
		$attributes,
		[
			'modalContent' => Components::render('video', $attributes),
			'modalId' => $buttonModalId,
			'modalCarouselHeader' => true
		]
	)
);

echo Components::classnames(['hello']);
```

And in my ruleset, I tried adding

```xml
<rule ref="WordPress.Security.EscapeOutput">
	<properties>
		<property name="customAutoEscapedFunctions" type="array">
			<element value="render" />
			<element value="Components::render" />
		</property>
	</properties>
</rule>
```

But this doesn't work. What does work (kinda) is:

```xml
<rule ref="WordPress.Security.EscapeOutput">
	<properties>
		<property name="customAutoEscapedFunctions" type="array">
			<element value="Components" />
		</property>
	</properties>
</rule>
```

I say kinda because only the `Components::render` should be on the auto escaped list, not `Components::classnames` (and the above will just see `Components` and mark it as safe).

Also, I think the related issue is #413.

With phpcsutils, how hard would be to add static calls to auto escaped list?

**Comment by dingo-d** [comment]
I kinda had to dig through the sniff and I think the issue is that the check for `autoEscapedFunctions` and `escapingFunctions` only happens inside the 

```if ( \T_STRING === $this->tokens[ $i ]['code'] ) {```

part of loop through echo'd components.

So the function name that is 'extracted' out inside this conditional only ever evaluates to the first string token found. In my case that is `Components`.

So a problem is a bit deeper, and could probably touch some other sniffs (as it usually ends up). We'd need to have a utility to recognize the static method calls. I think this could be added in the part of #764.

I could try to modify the existing sniff to handle static methods. Roughly it would be adding this in the loop

```php
if ($this->tokens[$i]['code'] === T_STRING) {
    $colon = $this->phpcsFile->findNext(Tokens::$emptyTokens, ($i + 1), null, true);
    
    if ($this->tokens[$colon]['code'] === T_DOUBLE_COLON) {
        $methodName = $this->phpcsFile->findNext(T_STRING, ($colon + 1));
    
        if ($this->tokens[$methodName]['code'] === T_STRING) {
        	// Static method call. 
        }
    }
}
```

Then, inside we could check the 

```php
if (
	$is_formatting_function
	|| isset( $this->autoEscapedFunctions[ $functionName ] )
	|| isset( $this->escapingFunctions[ $functionName ] )
) {
	continue;
}
```

Am I on the right track?

**Comment by jrfnl** [comment]
@dingo-d Good analysis, however IMO the bigger problem with this is not so much how to _detect_ these methods being used, but _how to allow them to be passed via a custom ruleset_ ?

Currently the property contains no keys and the function names as values.

If we would implement this ability to recognize methods, we would need to be able to distinguish between functions and methods to make sure that a method call using the same name as a global function from the "escaped list "doesn't cause a false negative and visa versa.

To that end, we'd need for methods to either have the class name or object name passed as well. The logical choice would be to add those as the array keys, however, a class could contain multiple escaping methods and if those would all use the same "key", then only the last method would actually be registered.

See the problem ?

**Comment by dingo-d** [comment]
Yeah, tricky. Could we add a new public property That would have a `name`, `type` (array) and a `value` that would denote the class? Then for every class the elements would just be all the escaped functions?

Something like

```xml
<rule ref="WordPress.Security.EscapeOutput">
	<properties>
		<property name="customAutoEscapedMethods" type="array" value="ClassName">
			<element value="custom_method_1"/>
			<element value="custom_method_2"/>
			<element value="custom_method_3"/>
		</property>
	</properties>
</rule>
```

Not sure if something like that would work 🤷🏼‍♂️ 

Then we'd have a public property `$customAutoEscapedMethods`, that would be

```php
$customAutoEscapedMethods = [
	'ClassName' => [
		'custom_method_1',
		'custom_method_2',
		'custom_method_3',
	]
]
```

Again, not sure if this is doable at all 🤷🏼‍♂️

**Comment by jrfnl** [comment]
@dingo-d 1) That is not an allowed format in PHPCS (the extra value in the `property` tag) and 2) that would only allow for methods in one (1) class to be added as a second set for a second class would overwrite the first.

I've been thinking along the lines of:
```xml
<rule ref="WordPress.Security.EscapeOutput">
	<properties>
		<property name="customAutoEscapedMethods" type="array">
			<element key="custom_method_1" value="ClassName"/>
			<element key="custom_method_2 value="ClassName""/>
			<element key="custom_method_3" value="$objectName"/>
		</property>
	</properties>
</rule>
```
Still fiddly and still wouldn't allow for multiple methods with the same name in different classes, but could become workable.

**Comment by dingo-d** [comment]
Yeah, so it's a limitation from the upstream. Are there any chances the configuration can be modified in v4 of phpcs?

**Comment by jrfnl** [comment]
> Yeah, so it's a limitation from the upstream. Are there any chances the configuration can be modified in v4 of phpcs?

Not really. More than anything, it's a limitation of the PHP array format which (rightfully) can not work with duplicate keys - or rather overwrites those.

**Comment by dingo-d** [comment]
What I thought was: is there a way to provide properties that will be parsed in the sniff as 

```php
$customAutoEscapedMethods = [
	'ClassName' => [
		'custom_method_1',
		'custom_method_2',
		'custom_method_3',
	],
	'OtherClassName' => [
		'custom_method_1',
		'custom_method_2',
		'custom_method_3',
	]
]
```

That way, every class is a separate key, and values are methods that are escaped (so nothing gets overwritten).

**Comment by jrfnl** [comment]
@dingo-d Well, we could make the value a comma-separated list of method names and then handle the exploding of that within the sniff. Does make it fiddly for ruleset configurators, but then again, it's advanced functionality.

```xml
<rule ref="WordPress.Security.EscapeOutput">
	<properties>
		<property name="customAutoEscapedMethods" type="array">
			<element key="ClassName" value="method_1,method2"/>
			<element key="OtherClassName" value="method_1,method_b"/>
		</property>
	</properties>
</rule>
```

**Comment by JPry** [comment]
> Well, we could make the value a comma-separated list of method names and then handle the exploding of that within the sniff. Does make it fiddly for ruleset configurators, but then again, it's advanced functionality.

Considering the limitations you've mentioned, I feel like this solution is still good in spite of its drawbacks. It is better than what we currently have.

I would vote in favor of something like this.

**Comment by janw-me** [comment]
In version `V2` it silenced the error when I used, in `V3` it shows the error again.
```xml
	<rule ref="WordPress.Security.EscapeOutput">
		<properties>
			<property name="customEscapingFunctions" type="array">
				<element value="Helpers"/>
			</property>
		</properties>
	</rule>
```

I've added the ignore to each. Or is there a better workaround?
 `//phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped`

**Comment by jrfnl** [comment]
@janw-me IIRC nothing changed regarding that property in v3, so to get a feel for what you are seeing we'd need more information. Also not sure if this is the right issue for your issue.

**Comment by janw-me** [comment]
I'll double check tomorrow.
But I'm pretty sure All I did was change the composer version, `composer update` and ran phpcs.

**Comment by janw-me** [comment]
@jrfnl I've recreated the bug in a test repo: https://github.com/janw-me/wpcs

I've tried to keep everything the same while strapping down the the unnecessary parts.
The problem might be in the .dist.

**Comment by dingo-d** [comment]
@janw-me can you check with the `dev-hotifx/escape-output-sniff` branch and see if it will work?

**Comment by janw-me** [comment]
@dingo-d It changes the output of the error to:
```
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 14 | ERROR | All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found
    |       | 'Helpers::wpcs_esc_str_rev'. (WordPress.Security.EscapeOutput.OutputNotEscaped)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Now it includes both the class and method, which is great :)
So from `found 'Helpers'.` to 'found `Helpers::wpcs_esc_str_rev'.`

I've tried the following extra xml:
```xml
    <rule ref="WordPress.Security.EscapeOutput">
        <properties>
            <property name="customEscapingFunctions" type="array">
                <element value="wpcs_esc_str_rev"/>
                <element value="Helpers"/>
                <element value="Helpers::wpcs_esc_str_rev"/>
                <element value="Helpers\:\:wpcs_esc_str_rev"/>
                <element key="Helpers" value="wpcs_esc_str_rev"/>
            </property>
        </properties>
    </rule>
```
I still get the error.

PS I'm available all afternoon (on slack) for testing.

**Comment by dingo-d** [comment]
I'll probably take a look tomorrow. Maybe there's something in the sniff I need to modify to take the static methods into account.

**Comment by dingo-d** [comment]
I think I found the issue, and will try to see how to solve it in my PR, I just need to get the feedback from the rest of the maintainers on the best way forward.

**Comment by dingo-d** [comment]
@janw-me I've updated the branch with some modifications, can you test it?

You should be able to pass the static methods with their respective classes now in the `customEscapingFunctions` property.

For instance, putting `\QualifiedExample\Namespaced\ClassName::statMethod` won't flag it as unsafe.

**Comment by janw-me** [comment]
@dingo-d  Yes, great this works :tada: 
I'll try updating the example in the wiki.

**Comment by dingo-d** [comment]
@janw-me please wait on updating the wiki, as the PR is not merged yet.

**Comment by Luc45** [comment]
@dingo-d The fix for **static** customEscapingFunctions is not in 3.0.1, it's still unreleased, right?

**Comment by dingo-d** [comment]
@Luc45 nope, the PR is still open, waiting for a review.

**Comment by Luc45** [comment]
@dingo-d Thanks for pointing me to that PR. I think it would work.

I believe that the issue that @janw-me and me are facing are caused by the `strtolower` added [here](https://github.com/WordPress/WordPress-Coding-Standards/commit/baf95127b9f9122f1beb077be2d1e83512ed78f3).

Essentially, I had to replace:

```xml
<property name="customEscapingFunctions" type="array">
    <element value="MyHelpers"/>
    <element value="my_custom_escaping_function"/>
</property>
```

With:

```xml
<property name="customEscapingFunctions" type="array">
    <element value="myhelpers"/>
    <element value="my_custom_escaping_function"/>
</property>
```

[check_code_is_escaped](https://github.com/WordPress/WordPress-Coding-Standards/blob/ec26b0e72dfedd1ba3e79b8d00066d0b93161a02/WordPress/Sniffs/Security/EscapeOutputSniff.php#L455) parses `MyHelpers::custom_sanitize( $output )`, as `MyHelpers`, `::`, `custom_sanitize`, all of which go through `strtolower`.

Your PR should work, as it lowercases all the `allEscapingFunctions`, and from what I understand makes it parse it as `MyHelpers::custom_sanitize`, which is nice and more intuitive for adding custom escaping/sanitizing functions.

**Comment by dero-newsuk** [comment]
@dingo-d @Luc45 Any chance this could still move forward?

**Comment by dingo-d** [comment]
I think you need to see with @rodrigoprimo and @jrfnl about this. I've made a PR but it wasn't being looked at so I've closed it (#2370).

I'm happy to reopen it if @rodrigoprimo wants to pick it up.

**Comment by dero-newsuk** [comment]
Thank you, @dingo-d.

There is a discrepancy now between [what the wiki says](https://github.com/WordPress/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#wordpresssecurityescapeoutput-custom-output-escaping-functions) and how the `customEscapingFunctions` configuration really works. Wiki claims support for static methods being a valid configuration option, but in reality that support has never been added.

It would be great to see the support added as it would allow us to get rid of trivial function wrappers calling static methods on our helper classes.

But at the very least the wiki should be updated to reflect reality.

Thoughts, @rodrigoprimo?

**Comment by rodrigoprimo** [comment]
Thanks for pointing out this discrepancy, @dero-newsuk. I updated the wiki to remove the line that mentioned passing a static method call to the list of custom escaping functions, as this is not currently supported. I will keep this issue in mind, and I will see if I can find time to work on it early next year.

**Comment by jrfnl** [comment]
> I'm happy to reopen it if @rodrigoprimo wants to pick it up.

@dingo-d I think that would be a good idea, but probably better once the "add support for PHPCS 4.0" project has been finished (to not delay it further). Ok ?
